doc: lay out board 'tiers' (#3191)
tock/tock: boards/README.md

The `/boards` directory contains the physical hardware platforms
that Tock supports.

Tock divides boards into three approximate 'tiers' of support.
These tiers are newly defined and are a bit informal as a result,
but the approximate definitions:

>  - **Tier 1:** The most feature-complete and thoroughly tested boards.
>                Loosely, these are boards used most regularly by core
>                team members or other highly engaged contributors.

Reviewer: ```suggestion
 - **Tier 1:** The most feature-complete and thoroughly tested boards. These
               are boards used most regularly by core team members or other
               highly engaged contributors. They are used as examples in the
               [Tock Book](https://book.tockos.org).
```

 - **Tier 2:** Platforms seeing reasonably regular use. These generally
               have broader, but still incomplete, peripheral support.
               They may also be 'relatives' of Tier 1 boards (e.g. a
               less-used varient in the nrf52 family) – likely in good
               shape, but not heavily tested. Some Tier 2 boards may
               have known issues, which are documented in release notes
               during release testing.
 - **Tier 3:** New or highly experimental. These should support the
               minimum platform requirements laid out in
               [the Porting documentation](../doc/Porting.md), but
               make no promises beyond that.

 - **Tier Z:** These are unstable _platforms_, where Tock bringup is
               happening in parallel with hardware development. These

>                are highly experimental by their nature.

Reviewer 2: ```suggestion
               are highly experimental by their nature, though some may actually 
               be better supported than Tier 2/3 platforms.
```

### Tier 1

| Board                                                             | Architecture     | MCU            | Interface  | App deployment              | QEMU Support? |
|-------------------------------------------------------------------|------------------|----------------|------------|-----------------------------|---------------|
| [Hail](hail/README.md)                                            | ARM Cortex-M4    | SAM4LC8BA      | Bootloader | tockloader                  | No            |
| [Imix](imix/README.md)                                            | ARM Cortex-M4    | SAM4LC8CA      | Bootloader | tockloader                  | No            |
| [Nordic nRF52840-DK](nordic/nrf52840dk/README.md)                 | ARM Cortex-M4    | nRF52840       | jLink      | tockloader                  | No            |

> | [Nordic nRF52840-Dongle](nordic/nrf52840_dongle/README.md)        | ARM Cortex-M4    | nRF52840       | jLink      | tockloader                  | No            |

Reviewer 3: Who uses the dongle regularly? I recall that we had some "issues" testing these, at least w.r.t. getting a TagConnect cable. I guess 1-2 maintainers have one? Might be considered Tier 2.

Reviewer: Agreed, I think part of being tier 1 is that the book is supported.

### Tier 2

| Board                                                             | Architecture     | MCU            | Interface  | App deployment              | QEMU Support? |
|-------------------------------------------------------------------|------------------|----------------|------------|-----------------------------|---------------|

> | [Nordic nRF52-DK](nordic/nrf52dk/README.md)                       | ARM Cortex-M4    | nRF52832       | jLink      | tockloader                  | No            |

Reviewer 3: I think the `nRF52DK` might easily be Tier 1.

Author: Does anyone regularly use this variant? I have a few, but I have abandoned them in favor of nrf52840dk's. If you're using plain 52's, I'm happy to promote it, but otherwise I think it should stay Tier 2 for the same reason the dongle is here – nrf family, very likely supported, but not continuously used / tested.

Reviewer 3: I was using that almost exclusively over the `nRF52840DK`, but I did not bring this board with me during my move, so starting virtually now I'll switch to the `nRF52840DK` as well I suppose. If no one else really uses that, that's good enough reasoning for me to keep it in Tier 2.

| [ACD52832](acd52832/README.md)                                    | ARM Cortex-M4    | nRF52832       | jLink      | tockloader                  | No            |
| [BBC Micro:bit v2](microbit_v2/README.md)                         | ARM Cortex-M4    | nRF52833       | openocd    | tockloader                  | No            |

> | [Nano 33 BLE](nano33ble/README.md)                                | ARM Cortex-M4    | nRF52840       | Bootloader | tockloader                  | No            |

Reviewer 2: I agree with Brad both of these should probably be tier 1

| [Clue nRF52840](clue_nrf52840/README.md)                          | ARM Cortex-M4    | nRF52840       | Bootloader | tockloader                  | No            |

> | [ST Nucleo F446RE](nucleo_f446re/README.md)                       | ARM Cortex-M4    | STM32F446      | openocd    | custom                      | #1827         |
> | [ST Nucleo F429ZI](nucleo_f429zi/README.md)                       | ARM Cortex-M4    | STM32F429      | openocd    | custom                      | #1827         |
> | [STM32F3Discovery kit](stm32f3discovery/README.md)                | ARM Cortex-M4    | STM32F303VCT6  | openocd    | custom                      | #1827         |
> | [STM32F412G Discovery kit](stm32f412gdiscovery/README.md)         | ARM Cortex-M4    | STM32F412G     | openocd    | custom                      | #1827         |

Reviewer 4: ```suggestion
| [ST Nucleo F446RE](nucleo_f446re/README.md)                       | ARM Cortex-M4    | STM32F446      | openocd    | custom                      | https://github.com/tock/tock/issues/1827 |
| [ST Nucleo F429ZI](nucleo_f429zi/README.md)                       | ARM Cortex-M4    | STM32F429      | openocd    | custom                      | https://github.com/tock/tock/issues/1827 |
| [STM32F3Discovery kit](stm32f3discovery/README.md)                | ARM Cortex-M4    | STM32F303VCT6  | openocd    | custom                      | https://github.com/tock/tock/issues/1827 |
| [STM32F412G Discovery kit](stm32f412gdiscovery/README.md)         | ARM Cortex-M4    | STM32F412G     | openocd    | custom                      | https://github.com/tock/tock/issues/1827 |
```

While we're in the file anyways, let's fix all the issue numbers that aren't properly linked. This will still become the text "#1827", but will have a link now.

| [Teensy 4.0](teensy40/README.md)                                  | ARM Cortex-M7    | i.MX RT 1062   | custom     | custom                      | No            |
| [Digilent Arty A-7 100T](arty_e21/README.md)                      | RISC-V RV32IMAC  | SiFive E21     | openocd    | tockloader                  | No            |

### Tier Z

| Board                                                             | Architecture     | MCU            | Interface  | App deployment              | QEMU Support? |
|-------------------------------------------------------------------|------------------|----------------|------------|-----------------------------|---------------|
| [QEMU RISC-V 32 bit `virt` platform](qemu_rv32_virt/README.md)    | RISC-V RV32IMAC  | QEMU           | custom     | custom                      | Yes (duh)     |
| [LiteX on Digilent Arty A-7](litex/arty/README.md)                | RISC-V RV32IMC   | LiteX+VexRiscV | custom     | tockloader (flash-file)[^1] | No            |
| [Verilated LiteX Simulation](litex/sim/README.md)                 | RISC-V RV32IMC   | LiteX+VexRiscv | custom     | tockloader (flash-file)[^1] | No            |

> | [OpenTitan Earlgrey on CW310](opentitan/earlgrey-cw310/README.md) | RISC-V RV32IMC   | EarlGrey       | custom     | custom                      | Yes (5.1)     |

Reviewer 5: This is one of the most used platforms and is part of the CI. I think it should be tier1 not experimental

Reviewer 3: I think that _experimental_ here really does mean that these board definitions are -- by their nature -- moving targets. As soon as there is a physical chip of OpenTitan with reasonable availability, I suppose that very few would be averse to classifying it as a Tier 1 board.

Nonetheless, the term "experimental" does sound like there's an expectation of instability associated with these boards. I might agree that OpenTitan is one of our more well-tested boards, and the `litex-sim-ci` workflow should make sure that the few features which the LiteX boards do support work quite reliably. I'm having a hard time coming up with a better name though.

Author: I tried splitting into tier Z1/Z2? I'm not sure I love the language or the binning I made though..

Reviewer 6: From the description of tier Z:

> Tock bringup is happening in parallel with hardware development

That perfectly describes Earlgrey right now. As long as you have to pin to a particular OT commit to make this port work, I don't think it should be tier 1.

Reviewer 3: > I tried splitting into tier Z1/Z2?

Did you push that? I'm not sure whether I'd be a fan of that. I think that for us, Tier 1/2/3/Z is already relatively granular, I fear with Z1/Z2 we'd get into pointless discussions about what it actually means to be more or less experimental. Is it stability of the hardware, stability of Tock given some pinned hardware revision, stability of the composite moving targets of Tock + hardware...?

Reviewer 4: It's here now if you look at the rendered view: https://github.com/tock/tock/blob/6b28bc6414e588c0d74dc50dea0568ae5d7ec56b/boards/README.md

I agree with your concern Leon, but I think practically it looks fine. Especially with the descriptions Pat put under them.

Reviewer 7: I suggest putting the clue_nrf52840 in tier 1 as it is fully supported and work with tockloader and tockbootloader. We still use the board for the display.

Author: Clue moved to tier 1.

Are we happy with the Z1/Z2 stuff? I'm open to whatever, but I think this is approaching best-first-cut that we can refine if/when it proves necessary?

Reviewer 3: > Are we happy with the Z1/Z2 stuff?

I'm not a fan for the reasons outlined above, but also not averse to this. I think it looks fine with the descriptions for Z1 & Z2.
